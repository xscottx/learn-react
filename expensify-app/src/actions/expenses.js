import uuid from 'uuid';
import database from '../firebase/firebase';

// Action creators creation

// Pure react-redux
// component calls action generator
// action generator returns object
// component dispatches object
// redux store changes

// Middleware style
// components calls action generator
// action generator returns function
// component dispatches function
// function runs (has the ability to dispatch other actions and do whatever it wants)

// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
  })

export const startAddExpense = (expenseData = {}) => {
  // this returns a function, have access to getState from thunk
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = '', 
      note = '', 
      amount = 0, 
      createdAt = 0
    } = expenseData;

    const expense = {description, note, amount, createdAt};

    // this is a function that returns a promise
    return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    })
  }
}

// REMOVE_EXPENSE
export const removeExpense = ({id} = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
})

// START_REMOVE_EXPENSE
export const startRemoveExpense = ({id} = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({id}));
    })
  }
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})

// START_EDIT_EXPENSE
export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates));
    })
  }
}

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

// START_SET_EXPENSES
export const startSetExpenses = () => {
  // this returns a function
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    // this is a function that returns a promise
    return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
        const expenses = [];
      
        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          })
        });
      
        dispatch(setExpenses(expenses));
      })
  }
}