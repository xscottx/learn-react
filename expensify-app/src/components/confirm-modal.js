import React from 'react';
import Modal from 'react-modal';

const ConfirmModal = (props) => (
  <Modal
    isOpen={!!props.selectedOption}
    contentLabel="Are you sure?"
    onRequestClose={props.handleCancel}
    ariaHideApp={false}
    closeTimeoutMS={200}
    className="modal"
  >
    <h3 className="modal__title">Delete expense?</h3>
    {props.selectedOption && <p className="modal__body">{props.selectedOption}</p>}
    <button className="button button--secondary" onClick={props.handleRemoveOption}>Delete</button>
  </Modal>
);

export default ConfirmModal;