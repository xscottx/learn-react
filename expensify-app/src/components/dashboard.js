import React from 'react';
import ExpenseList from './expenses-list';
import ExpenseListFilters from './expense-list-filters';

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
)

export default ExpenseDashboardPage;