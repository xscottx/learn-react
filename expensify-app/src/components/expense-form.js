import React from 'react';
import moment from 'moment'; // http://momentjs.com/docs/#/displaying/
import {SingleDatePicker} from 'react-dates'; // https://github.com/airbnb/react-dates
import {Link} from 'react-router-dom';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
   
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    }
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({description}));
  }

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({note}));
  }

  onAmountChange = (e) => {
    const amount = e.target.value;
    // https://regex101.com/
    if (!amount || amount.match(/^\d{1,}(\.?\d{0,2})$/)) {
      this.setState(() => ({amount}));
    }
  }

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  }

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  }

  onSubmit = (e) => {
    e.preventDefault(); // will prevent browser from doing full page refresh

    if (!this.state.description || !this.state.amount) {
      // Set error state equal to 'Please provide description and amount'
      this.setState(() => ({ error: 'Please provide description and amount'}));
    }
    else {
      // Clear the error 
      this.setState(() => ({ error: '' }));
      // instead of doing work inside expenseform, we fire props.onSubmit to allow AddExpense & EditExpense to use
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      })
    }
  }

  render() {
    return (
      <div>
        <div className="button__container">
          <Link to="/dashboard" className="button button--form-link">Go Back</Link>
        </div>
        <form className="form" onSubmit={this.onSubmit} >
          {this.state.error && <p className="form__error">{this.state.error}</p>}
          <input 
            className="text-input"
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input 
            className="text-input"
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false }
            block
          />
          <textarea
            className="textarea"
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          >
          </textarea>
          <div>
            <button className="button">
              Save Expense
            </button>
          </div>
        </form>
      </div>
    )
  }
}