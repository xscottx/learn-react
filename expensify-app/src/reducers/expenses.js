// Reducers creation
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      // spread operator used to return a new collection without modifying existing state, this appends expense to array
      return [...state, action.expense];
    case 'REMOVE_EXPENSE':
      // filter out object array inside expenses that do not match the expense id
      return state.filter(({id}) => id !== action.id);
    case 'EDIT_EXPENSE':
      // map will allow you to traverse thru array, and pass in an 'updater' function to make changes if needed
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates
          }
        }
      })
    default:
      return state;
  }
}