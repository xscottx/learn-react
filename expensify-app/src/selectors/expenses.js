import moment from 'moment';

// Get visible expenses
// JS toLowerCase() lowers strings, and .includes() checks if the text is inside string
const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {

  return expenses.filter((expense) => {
    // create a moment for expense and filter on dates before or after the start/end date match.
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    }
    else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1;
    }
    
  })
}

export default getVisibleExpenses;