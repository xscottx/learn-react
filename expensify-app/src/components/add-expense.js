import React from 'react';
import ExpenseForm from './expense-form';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';

// avoid in-line functions by moving over to class based components
export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/dashboard');
  }

  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm 
          onSubmit={this.onSubmit}
        />
      </div>
    )
  }
}

// const AddExpensePage = (props) => (
//   <div>
//     <h1>Add Expense</h1>
//     <ExpenseForm 
//       onSubmit={(expense) => {
//         // leverage props.dispatch from redux to dispatch an action creator (addExpense)
//         // props.dispatch(addExpense(expense));
//         props.onSubmit(expense);
//         // leverage the history props from redux
//         props.history.push('/');
//       }}
//     />
//   </div>
// )

// New way - function 
const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
})

// Old way - function that returns an object, can reduce to above to implicitly return object
// const mapDispatchToProps = (dispatch) => {
//   return {
//     onSubmit: (expense) => dispatch(addExpense());
//   }
// }

// need to use connect to get access to props.dispatch (HOC)
export default connect(undefined, mapDispatchToProps)(AddExpensePage);