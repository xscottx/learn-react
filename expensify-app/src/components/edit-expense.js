import React from 'react';
import ExpenseForm from './expense-form';
import {connect} from 'react-redux';
import {startEditExpense, startRemoveExpense} from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/dashboard');
  }

  onClick = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/dashboard');
  }

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm 
            className="form"
            expense={this.props.expense}
            onSubmit={this.onSubmit}
          />
          <button className="button button--secondary" onClick={this.onClick}>Remove Expense</button>
        </div>
      </div>
    )
  }
};

// const EditExpensePage = (props) => {
//   return (
//     <div>
//       <ExpenseForm 
//         expense={props.expense}
//         onSubmit={(expense) => {
//           props.dispatch(editExpense(props.expense.id, expense));
//           props.history.push('/');
//         }}
//       />
//       <button onClick={() => {
//         props.dispatch(removeExpense({ id: props.expense.id }));
//         props.history.push('/');
//       }}>Remove</button>
//     </div>
//   )
// }

// need to set up mapStateToProps because we want to pass values to component, have access to props in the 2nd argument
// adding expense prop to HOC
// leverage find to find a single item if the id matches
// if an expression is being returned can just use shorthand arrow function to return the expression
//  old school:  (expense) => { return return expense.id === props.match.params.id; }
const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense: (id, updates) => dispatch(startEditExpense(id, updates)),
  startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);