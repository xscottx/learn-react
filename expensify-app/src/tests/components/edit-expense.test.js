import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/edit-expense';
import expenses from '../fixtures/expenses';

let editExpense, history, removeExpense, wrapper;

beforeEach(() => {
  editExpense = jest.fn(); // spy
  history = { push: jest.fn() }; // spy
  removeExpense = jest.fn(); // spy
  wrapper = shallow(
    <EditExpensePage 
      editExpense={editExpense} 
      history={history} 
      removeExpense={removeExpense} 
      expense={expenses[0]}
    />);
})

test('should render edit-expense', () => {
  expect(wrapper).toMatchSnapshot();  
})

test('should handle edit expense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
})

test('should handle remove expense', () => {
  wrapper.find('button').simulate('click'); // use simulate click here instead of onSubmit
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(removeExpense).toHaveBeenLastCalledWith({id: expenses[0].id});
})