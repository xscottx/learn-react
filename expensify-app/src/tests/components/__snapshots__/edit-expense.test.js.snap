// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit-expense 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "Gum",
        "id": "1",
        "note": "yummy",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
